### PR TITLE
Check DNS on replication loop start not at task start

### LINKED ISSUE #3188: Handle replication DNS resolution in Kubernetes
### I did this

Deployed multiple replicas in a StatefulSet in a Kubernetes environment. I used the [Pod subdomain](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id) for referring to each replica.

#### Configuration example

Pod 0:
```toml
[replication]
origin = "repl://test-change-kanidm-replicas-default-0.test-change-kanidm-replicas:8444"
bindaddress = "0.0.0.0:8444"

[replication."repl://test-change-kanidm-replicas-default-1.test-change-kanidm-replicas:8444"]
type = "mutual-pull"
partner_cert = "****"
automatic_refresh = false
```

Pod 1:
```toml
[replication]
origin = "repl://test-change-kanidm-replicas-default-1.test-change-kanidm-replicas:8444"
bindaddress = "0.0.0.0:8444"

[replication."repl://test-change-kanidm-replicas-default-0.test-change-kanidm-replicas:8444"]
type = "mutual-pull"
partner_cert = "****"
automatic_refresh = false
```

### I expected the following

Replication works bidirectionally between different pods.

### This happened instead

Replicas try to connect to the IP resolved when they start. So, just the last restarted replica will work because pods get a different IP when recreated.

### Kanidm version details

- Output of `kanidm(d) version`:  `kanidmd 1.4.2`
- Are you running it in a container? If so, which image/tag?: Yes, kanidm/server:latest

### Any other comments

You can workaround it by forcing a restart of the process but it is not the intended way of working in Kubernetes and when a node is down or you just have to change some configuration, Pod will be recreated.

### PR DESCRIPTION
**Pull request by yaleman**
# Change summary

- instead of checking DNS for the replication host on startup, do it when we're about to replicate

Fixes #3188

Checklist

- [ ] This PR contains no AI generated code
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

**Comment by yaleman** [comment]
@jinnatar any chance you've got a platform that'd  allow live-testing of this one? (Or @pando85 since it's your bug 😄)

**Comment by pando85** [comment]
Yes, of course. It will be easier to test it with a Docker image, but I can also create one.

I don't know if we could improve the logic in the future and re-use the connection if possible. And if it fails then create a new socket, resolve the DNS, and so on.

Thank you for the fix @yaleman . I will let you know about the test results.

**Comment by pando85** [comment]
I can confirm that it is working.

For reproducible steps:
```bash
CONTAINER_BUILD_ARGS=--load make build/kanidmd
git clone --branch fix/remove-pod-service-workaround \
  --single-branch --depth=1 https://github.com/pando85/kaniop
cd kaniop
make e2e
kind load --name chart-testing docker-image kanidm/server:devel
cat << EOF > /tmp/kanidm.yaml
apiVersion: v1
kind: Secret
metadata:
  name: my-idm-tls
type: kubernetes.io/tls
data:
  tls.crt: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUNoRENDQWl1Z0F3SUJBZ0lCQWpBS0JnZ3Foa2pPUFFRREFqQ0JoREVMTUFrR0ExVUVCaE1DUVZVeEREQUsKQmdOVkJBZ01BMUZNUkRFUE1BMEdBMVVFQ2d3R1MyRnVhV1J0TVJ3d0dnWURWUVFEREJOTFlXNXBaRzBnUjJWdQpaWEpoZEdWa0lFTkJNVGd3TmdZRFZRUUxEQzlFWlhabGJHOXdiV1Z1ZENCaGJtUWdSWFpoYkhWaGRHbHZiaUF0CklFNVBWQ0JHVDFJZ1VGSlBSRlZEVkVsUFRqQWVGdzB5TkRFd01UTXlNRFF6TWpoYUZ3MHlOREV3TVRneU1EUXoKTWpoYU1JR0FNUXN3Q1FZRFZRUUdFd0pCVlRFTU1Bb0dBMVVFQ0F3RFVVeEVNUTh3RFFZRFZRUUtEQVpMWVc1cApaRzB4R0RBV0JnTlZCQU1NRDJsa2JTNWxlR0Z0Y0d4bExtTnZiVEU0TURZR0ExVUVDd3d2UkdWMlpXeHZjRzFsCmJuUWdZVzVrSUVWMllXeDFZWFJwYjI0Z0xTQk9UMVFnUms5U0lGQlNUMFJWUTFSSlQwNHdXVEFUQmdjcWhrak8KUFFJQkJnZ3Foa2pPUFFNQkJ3TkNBQVJUaTdocW8wWjNCVTNwOTV6NmhRelBtWUFveDNiS2ZBQXU0R21ZOFFoZgpCcTNUTThoZi8vRVBjU1FtYm1xRlVkc3BJMHIzMWhmYzBsSVhIWDVxTkJhSW80R1BNSUdNTUFrR0ExVWRFd1FDCk1BQXdEZ1lEVlIwUEFRSC9CQVFEQWdXZ01CTUdBMVVkSlFRTU1Bb0dDQ3NHQVFVRkJ3TUJNQjBHQTFVZERnUVcKQkJRYWFySFRSbTRZajZUTVB6dmR1QUI3bk9ES0h6QWZCZ05WSFNNRUdEQVdnQlRhT2FQdVhtdExEVEpWdisrVgpZQmlRcjlnSENUQWFCZ05WSFJFRUV6QVJnZzlwWkcwdVpYaGhiWEJzWlM1amIyMHdDZ1lJS29aSXpqMEVBd0lEClJ3QXdSQUlnUXBMczlNWnZCUlVwUjE1d3ZTd0lxL1F5V290dlZnLzN2Wmw4RDFtVEZ6OENJRVZibSsvK3o0SkwKTFl3TlhuZXJ2OU5jK2FuR3R6KzliZVQ0YmtTNENwSlMKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQotLS0tLUJFR0lOIENFUlRJRklDQVRFLS0tLS0KTUlJQ1BqQ0NBZVNnQXdJQkFnSUJBVEFLQmdncWhrak9QUVFEQWpDQmhERUxNQWtHQTFVRUJoTUNRVlV4RERBSwpCZ05WQkFnTUExRk1SREVQTUEwR0ExVUVDZ3dHUzJGdWFXUnRNUnd3R2dZRFZRUUREQk5MWVc1cFpHMGdSMlZ1ClpYSmhkR1ZrSUVOQk1UZ3dOZ1lEVlFRTERDOUVaWFpsYkc5d2JXVnVkQ0JoYm1RZ1JYWmhiSFZoZEdsdmJpQXQKSUU1UFZDQkdUMUlnVUZKUFJGVkRWRWxQVGpBZUZ3MHlOREV3TVRNeU1EUXpNamhhRncweU5ERXhNVEl5TURRegpNamhhTUlHRU1Rc3dDUVlEVlFRR0V3SkJWVEVNTUFvR0ExVUVDQXdEVVV4RU1ROHdEUVlEVlFRS0RBWkxZVzVwClpHMHhIREFhQmdOVkJBTU1FMHRoYm1sa2JTQkhaVzVsY21GMFpXUWdRMEV4T0RBMkJnTlZCQXNNTDBSbGRtVnMKYjNCdFpXNTBJR0Z1WkNCRmRtRnNkV0YwYVc5dUlDMGdUazlVSUVaUFVpQlFVazlFVlVOVVNVOU9NRmt3RXdZSApLb1pJemowQ0FRWUlLb1pJemowREFRY0RRZ0FFaXo1bXFIb3pwc2o1aUdDREg4dVNKeThURnFOSUduSXc4VS9MCnN3eWVGVEdIVDRTMkh3QmI3UUFvdVlWdVhkd0w4aFpHTXR6QXFvWU1GaEN0MWVwWGpxTkZNRU13RWdZRFZSMFQKQVFIL0JBZ3dCZ0VCL3dJQkFEQU9CZ05WSFE4QkFmOEVCQU1DQVFZd0hRWURWUjBPQkJZRUZObzVvKzVlYTBzTgpNbFcvNzVWZ0dKQ3YyQWNKTUFvR0NDcUdTTTQ5QkFNQ0EwZ0FNRVVDSUd5WmpCczRwcDFIQWxGZGswbWRWQno0CjQ0MHQ4cFJIaDgvU09ZNVp0TWNTQWlFQTZxT2Y5YVFiV3dFWExqMGphalg5bEhnZHFsd1JrN3dubnlMTUdGNS8KbHo4PQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
  tls.key: LS0tLS1CRUdJTiBQUklWQVRFIEtFWS0tLS0tCk1JR0hBZ0VBTUJNR0J5cUdTTTQ5QWdFR0NDcUdTTTQ5QXdFSEJHMHdhd0lCQVFRZ0hMWm9NVFVNYWR4T0tNbHQKVHEva0RudU4zOEdDSmt3ajhZMmtxeUdsY2YraFJBTkNBQVJUaTdocW8wWjNCVTNwOTV6NmhRelBtWUFveDNiSwpmQUF1NEdtWThRaGZCcTNUTThoZi8vRVBjU1FtYm1xRlVkc3BJMHIzMWhmYzBsSVhIWDVxTkJhSQotLS0tLUVORCBQUklWQVRFIEtFWS0tLS0tCg==
---
apiVersion: kaniop.rs/v1beta1
kind: Kanidm
metadata:
  name: my-idm
spec:
  domain: idm.example.com
  image: kanidm/server:devel
  replicaGroups:
    - name: default
      replicas: 3
      primaryNode: true
  logLevel: info
  storage:
    volumeClaimTemplate:
      apiVersion: v1
      kind: PersistentVolumeClaim
      metadata: {}
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 1Gi
EOF
kubectl apply -f /tmp/kanidm.yaml
```

**Comment by yaleman** [comment]
I'm going to have to check out that operator, thanks for testing the fix out!

**Comment by pando85** [comment]
It's in an early stage I will announce it once that's more stable and have some docs.

**Comment by yaleman** [comment]
Cool, will wait for @Firstyear to sign off

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: server/core/src/repl/mod.rs [hunk 1]
```diff
@@ -465,8 +457,17 @@ async fn repl_task(
...
 
-    // Okay, all the parameters are setup. Now we wait on our interval.
+    // Okay, all the parameters are set up. Now we wait on our interval.
     loop {
+        // we resolve the DNS entry to the ip:port each time we attempt a connection to avoid stale DNS issues, ref #3188
+        let socket_addrs = match origin.socket_addrs(|| Some(443)) {
+            Ok(sa) => sa,
+            Err(err) => {
+                error!(?err, "Replica origin could not resolve to ip:port");
+                return;
```
[the remark this answers is not in the document]
Comment (yaleman) [reply server/core/src/repl/mod.rs:467]: Good catch, thanks

**Review by Firstyear — APPROVED** [review]

**Pull request merged**